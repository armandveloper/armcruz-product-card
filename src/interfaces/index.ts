import { Props as ProductCardProps } from '../components/ProductCard';
import { Props as ProductTitleProps } from '../components/ProductTitle';
import { Props as ProductImageProps } from '../components/ProductImage';
import { Props as ProductButtonsProps } from '../components/ProductButtons';

export interface ProductContextProps extends ProductCounter {
	product: Product;
	isMaxCountReached: boolean;
}

export interface Product {
	id: string;
	img?: string;
	title: string;
}

export interface ProductCounter {
	counter: number;
	isMaxCountReached: boolean;
	maxCount?: number;
	increaseBy: (value: number) => void;
	reset: () => void;
}

export interface ProductCardHOCProps {
	(props: ProductCardProps): JSX.Element;
	Title: (props: ProductTitleProps) => JSX.Element;
	Image: (props: ProductImageProps) => JSX.Element;
	Buttons: (props: ProductButtonsProps) => JSX.Element;
}

export interface OnChangeArgs {
	product: Product;
	count: number;
}

export interface CartState {
	[key: string]: ProductInCart;
}

export interface ProductInCart extends Product {
	count: number;
}

export interface IInitialValues {
	count?: number;
	maxCount?: number;
}

export interface ProductCardHandlers {
	count: number;
	isMaxCountReached: boolean;
	maxCount?: number;
	product: Product;

	increaseBy: (value: number) => void;
	reset: () => void;
}
