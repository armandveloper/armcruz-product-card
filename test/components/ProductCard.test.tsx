import * as React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src';
import { product } from '../data';

const { act } = renderer;

describe('ProductCard', () => {
  test('can render the component', () => {
    const wrapper = renderer.create(
      <ProductCard product={product}>{() => <h1>Product Card</h1>}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('can increase the counter', () => {
    const wrapper = renderer.create(
      <ProductCard product={product}>
        {({ count, increaseBy }) => {
          return (
            <>
              <h1>Product Card</h1>
              <span>{count}</span>
              <button onClick={() => increaseBy(1)}>+1</button>
            </>
          );
        }}
      </ProductCard>
    );

    let tree = wrapper.toJSON();

    act(() => {
      (tree as any).children[2].props.onClick();
    });

    tree = wrapper.toJSON();

    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
