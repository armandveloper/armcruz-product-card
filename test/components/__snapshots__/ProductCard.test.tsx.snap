// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard can increase the counter 1`] = `
<div
  className="productCard "
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    +1
  </button>
</div>
`;

exports[`ProductCard can render the component 1`] = `
<div
  className="productCard "
>
  <h1>
    Product Card
  </h1>
</div>
`;
